contracts/rollover/CrossCurrencyRollover.sol

```diff
@@ -0,0 +1,428 @@
+                amountIn: amountIn,
+                amountOutMinimum: amountOutMinimum,
+                sqrtPriceLimitX96: 0
+            });
+
+        // execute the swap
+        amountOut = swapRouter.exactInputSingle(params);
+    }
+
+    /**
+     * @notice Collects payment funds and performs currency swap for cross-currency loan repayment.
+     *         Any funds that are not covered by closing out the old loan must be covered by
+     *         the borrower.  Any excess funds are sent to the borrower.
+     *
+     * @param borrower               The address of the borrower.
+     *                               executing the repayment and swap.
+     * @param lender                 The address of the new lender.
+     * @param swapParams             The parameters for the currency swap.
+     * @param newLoanTerms           The terms of the new loan.
+     * @param oldLoanData            The loan data of the original loan.
+     * @param oldLoanId              The ID of the original loan to be rolled over.
+     *
+     * @return repayAmount           The amount to be repaid to the old loan.
+     * @return feeSnapshot           Principal and interest fees pertaining to rollover.
+     */
+    function _processSettlement(
+        address borrower,
+        address lender,
+        OriginationLibrary.SwapParameters memory swapParams,
+        LoanLibrary.LoanTerms memory newLoanTerms,
+        LoanLibrary.LoanData memory oldLoanData,
+        uint256 oldLoanId
+    ) internal nonReentrant returns (uint256 repayAmount, LoanLibrary.FeeSnapshot memory feeSnapshot){
+        // pull funds from the new lender
+        IERC20(newLoanTerms.payableCurrency).safeTransferFrom(
+            lender,
+            address(this),
+            newLoanTerms.principal
+        );
+
+        // swap the new lender payment from new currency to the original currency
+        // the swap will revert if swappedAmount < minAmountOut
+        uint256 swappedAmount = _swapExactInputSingle(
+            newLoanTerms.payableCurrency,
+            oldLoanData.terms.payableCurrency,
+            newLoanTerms.principal,
+            swapParams.minAmountOut,
+            swapParams.poolFeeTier,
+            address(this)
+        );
+
+        address oldLender = ILoanCore(loanCore).lenderNote().ownerOf(oldLoanId);
+
+        // calculate settle amounts
+        // get interest amount due
+        uint256 interestAmount = InterestCalculator.getProratedInterestAmount(
+            oldLoanData.balance,
+            oldLoanData.terms.interestRate,
+            oldLoanData.terms.durationSecs,
+            oldLoanData.startDate,
+            oldLoanData.lastAccrualTimestamp,
+            block.timestamp
+        );
+
+        // calculate the repay amount to settle the original loan
+        repayAmount = oldLoanData.terms.principal + interestAmount;
+
+        LoanLibrary.FeeSnapshot memory feeSnapshot = LoanLibrary.FeeSnapshot({
```

Comment:
I dont think we need to create new feeSnapshot in memory here. Can use `oldLoanData.lenderInterestFee` and `oldLoanData.lenderPrincipalFee` in the rolloverAmounts call. Also seems best to remove it from here and `_executeRollover()`. Create it in `_initializeRolloverLoan()` since that is the only place it is needed. and your loading the `oldLoanData` there anyways. Will save some gas

Reply:
Good suggestion.